contracts/eosio.system/src/finalizer_key.cpp

```diff
@@ -0,0 +1,252 @@
+         k.id            = _finalizer_keys.available_primary_key();
+         k.finalizer     = finalizer;
+         k.finalizer_key = finalizer_key;
+      });
+
+      // Update finalizers table
+      auto f_itr = _finalizers.find(finalizer.value);
+      if( f_itr == _finalizers.end() ) {
+         // First time the finalizer to register a finalier key,
+         // make the key active
+         _finalizers.emplace( finalizer, [&]( auto& f ) {
+            f.finalizer           = finalizer;
+            f.active_key_id       = key_itr->id;
+            f.active_key          = { fin_key_g1.begin(), fin_key_g1.end() };
+            f.num_registered_keys = 1;
+         });
+      } else {
+         // Update num_registered_keys
+         _finalizers.modify( f_itr, same_payer, [&]( auto& f ) {
+            ++f.num_registered_keys;
+         });
+      }
+   }
+
+   // Action to activate a finalizer key
+   void system_contract::actfinkey( const name& finalizer, const std::string& finalizer_key ) {
+      require_auth( finalizer );
+
+      auto producer = _producers.find( finalizer.value );
+      check( producer != _producers.end(), "finalizer is not a registered producer");
+
+      // Check the key is registered
+      const auto fin_key_g1 = eosio::decode_bls_public_key_to_g1(finalizer_key);
+      auto idx = _finalizer_keys.get_index<"byfinkey"_n>();
+      auto hash = eosio::sha256(fin_key_g1.data(), fin_key_g1.size());
+      auto fin_key = idx.lower_bound(hash);
+      check(fin_key != idx.end(), "finalizer key was not registered");
+
+      // Check the key belongs to finalizer
+      check(fin_key->finalizer == finalizer, "finalizer_key was not registered by the finalizer");
+
+      // Check finalizer itself
+      auto fin = _finalizers.find(finalizer.value);
+      check( fin != _finalizers.end(), "finalizer is not in the finalizers table" );
+      check( fin->num_registered_keys > 0, "num_registered_keys of the finalizer must be greater than one" );
+
+      // Check if the finalizer key is not already active
+      check( fin_key->id != fin->active_key_id, "the finalizer key was already active" );
+
+      auto old_active_key_id = fin->active_key_id;
+
+      _finalizers.modify( fin, same_payer, [&]( auto& f ) {
+         f.active_key_id = fin_key->id;
+         f.active_key    = { fin_key_g1.begin(), fin_key_g1.end() };
+      });
+
+      // Replace the finalizer policy immediately if the finalizer is
+      // participating in current voting
+      if( _gstate5.last_finalizer_key_ids.contains(old_active_key_id) ) {
+         replace_key_in_finalizer_policy(finalizer, old_active_key_id, fin_key->id);
+      }
+   }
+
+   // replace the key in last finalizer policy and call set_finalizers host function immediately
+   void system_contract::replace_key_in_finalizer_policy(const name& finalizer, uint64_t old_id, uint64_t new_id) {
+      std::set<uint64_t> next_finalizer_key_ids {_gstate5.last_finalizer_key_ids};
```

Comment:
I think this should be a table. But if kept as a set, I don't see a reason to copy it here.